internal/server/evaluation/evaluator.go

```diff
@@ -275,6 +200,83 @@ func (e *Evaluator) Evaluate(ctx context.Context, r *flipt.EvaluationRequest) (r
 	return resp, nil
 }
 
+// doConstraintsMatch is a utility function that will return if all or any constraints have matched for a segment depending
+// on the match type.
+func doConstraintsMatch(logger *zap.Logger, evalCtx map[string]string, constraints []storage.EvaluationConstraint, segmentMatchType flipt.MatchType) (bool, error) {
+	constraintMatches := 0
+
+	for _, c := range constraints {
+		v := evalCtx[c.Property]
+
+		var (
+			match bool
+			err   error
+		)
+
+		switch c.Type {
+		case flipt.ComparisonType_STRING_COMPARISON_TYPE:
+			match = matchesString(c, v)
+		case flipt.ComparisonType_NUMBER_COMPARISON_TYPE:
+			match, err = matchesNumber(c, v)
+		case flipt.ComparisonType_BOOLEAN_COMPARISON_TYPE:
+			match, err = matchesBool(c, v)
+		case flipt.ComparisonType_DATETIME_COMPARISON_TYPE:
+			match, err = matchesDateTime(c, v)
+		default:
+			return false, errs.ErrInvalid("unknown constraint type")
+		}
+
+		if err != nil {
+			return false, err
+		}
+
+		if match {
+			// increase the matchCount
+			constraintMatches++
+
+			switch segmentMatchType {
+			case flipt.MatchType_ANY_MATCH_TYPE:
+				// can short circuit here since we had at least one match
+				break
+			default:
+				// keep looping as we need to match all constraints
+				continue
+			}
+		} else {
+			// no match
+			switch segmentMatchType {
+			case flipt.MatchType_ALL_MATCH_TYPE:
+				// we can short circuit because we must match all constraints
+				break
+			default:
+				// keep looping to see if we match the next constraint
+				continue
+			}
+		}
+	}
+
+	var matched = true
+
+	switch segmentMatchType {
+	case flipt.MatchType_ALL_MATCH_TYPE:
+		if len(constraints) != constraintMatches {
+			logger.Debug("did not match ALL constraints")
+			matched = false
+		}
+
+	case flipt.MatchType_ANY_MATCH_TYPE:
+		if len(constraints) > 0 && constraintMatches == 0 {
+			logger.Debug("did not match ANY constraints")
+			matched = false
+		}
+	default:
+		logger.Error("unknown match type", zap.Int32("match_type", int32(segmentMatchType)))
```

Comment:
this would be a breaking change potentially.. but I'm wondering if we should throw an error here if the `segment.MatchType` isnt' valid. similar to how we do above for `constraintType`: https://github.com/flipt-io/flipt/pull/1806/files#diff-30fffd9a6749cb98b815b799960ed922ffff95f24feeaac79203eb0ee7d18167R226

Reply:
I am down to implement this. Whatever decision we make, but yeah definitely a breaking change.